Error message if `libffi` is not present

I get the following error if I do not have libffi present.  Apparently @lpw25's error message is much more friendly and helpful :)

I'm on a Mac and using homebrew.

```text
==== ERROR [while installing ctypes.0.1] ====
# opam-version    1.0.1
# os              darwin
# command         make
# path            /Users/amir/.opam/4.01.0dev+trunk/build/ctypes.0.1
# compiler        4.01.0dev+trunk
# exit-code       2
# env-file        /Users/amir/.opam/4.01.0dev+trunk/build/ctypes.0.1/ctypes-18112-ad8886.env
# stdout-file     /Users/amir/.opam/4.01.0dev+trunk/build/ctypes.0.1/ctypes-18112-ad8886.out
# stderr-file     /Users/amir/.opam/4.01.0dev+trunk/build/ctypes.0.1/ctypes-18112-ad8886.err
### stdout ###
oasis setup
ocaml setup.ml -configure --enable-tests
testing for brew: .............................. available
testing for pkg-config: ........................ available
testing for libffi:
### stderr ###
Exception: End_of_file.
E: Failure("Command 'ocaml discover.ml -ocamlc /Users/amir/.opam/4.01.0dev+trunk/bin/ocamlc.opt -ext-obj .o -exec-name a.out' terminated with error code 2")
make: *** [setup.data] Error 1

'opam install ctypes' failed.
```

--- a/.gitignore
+++ b/.gitignore
@@ -1,5 +1,6 @@
 _build
-setup.data
+libffi.config
+asneeded.config
 src/ctypes/ctypes_primitives.ml
 src/ctypes_config.h
 src/ctypes_config.ml

--- a/Makefile
+++ b/Makefile
@@ -11,7 +11,7 @@ BASE_PROJECTS=configure libffi-abigen configured ctypes ctypes-top
 FOREIGN_PROJECTS=ctypes-foreign-base ctypes-foreign-threaded ctypes-foreign-unthreaded
 STUB_PROJECTS=cstubs
 PROJECTS=$(BASE_PROJECTS) $(FOREIGN_PROJECTS) $(STUB_PROJECTS)
-GENERATED=src/ctypes_config.h src/ctypes_config.ml setup.data src/ctypes/ctypes_primitives.ml src/ctypes-foreign-base/dl.ml src/ctypes-foreign-base/dl_stubs.c
+GENERATED=src/ctypes_config.h src/ctypes_config.ml libffi.config src/ctypes/ctypes_primitives.ml src/ctypes-foreign-base/dl.ml src/ctypes-foreign-base/dl_stubs.c
 OCAML_FFI_INCOPTS=$(libffi_opt)
 export CFLAGS DEBUG
 
@@ -25,7 +25,7 @@ OS_ALT_SUFFIX=.unix
 endif
 
 # public targets
-all: setup.data $(PROJECTS)
+all: libffi.config $(PROJECTS)
 
 ctypes-base: $(BASE_PROJECTS)
 ctypes-foreign: ctypes-base $(FOREIGN_PROJECTS)
@@ -144,9 +144,11 @@ src/ctypes/ctypes_primitives.ml: $(BUILDDIR)/configure.native
 src/ctypes-foreign-base/libffi_abi.ml: $(BUILDDIR)/libffi-abigen.native
 	$< > $@
 
-setup.data: src/discover/commands.mli src/discover/commands.ml src/discover/discover.ml
+libffi.config: src/discover/commands.mli src/discover/commands.ml src/discover/discover.ml
 	@ocamlfind ocamlc -o discover -package str,bytes -linkpkg $^ -I src/discover
 	./discover -ocamlc "$(OCAMLFIND) ocamlc" > $@ || (rm $@ && false)
+
+asneeded.config:
 	./src/discover/determine_as_needed_flags.sh >> $@
 
 # dependencies
@@ -176,5 +178,5 @@ uninstall:
 .PHONY: depend distclean clean configure all install $(PROJECTS)
 
 include .depend Makefile.rules Makefile.examples Makefile.tests
--include setup.data
-
+-include libffi.config
+-include asneeded.config

--- a/Makefile
+++ b/Makefile
@@ -167,7 +167,7 @@ META-install:
 install-%: PROJECT=$*
 install-%:
 	$(if $(filter yes,$($(PROJECT).install)),\
-		$(OCAMLFIND) install -add ctypes $^ \
+		$(OCAMLFIND) install -add ctypes -optional $^ \
                    $(LIB_TARGETS) $(LIB_TARGET_EXTRAS) \
                    $(INSTALL_MLIS) $(INSTALL_CMIS) \
                    $(INSTALL_HEADERS) \

--- a/Makefile
+++ b/Makefile
@@ -7,7 +7,10 @@ OCAMLDEP=$(OCAMLFIND) ocamldep
 OCAMLMKLIB=$(OCAMLFIND) ocamlmklib
 VPATH=src examples
 BUILDDIR=_build
-PROJECTS=configure libffi-abigen configured ctypes cstubs ctypes-foreign-base ctypes-foreign-threaded ctypes-foreign-unthreaded ctypes-top
+BASE_PROJECTS=configure libffi-abigen configured ctypes ctypes-top
+FOREIGN_PROJECTS=ctypes-foreign-base ctypes-foreign-threaded ctypes-foreign-unthreaded
+STUB_PROJECTS=cstubs
+PROJECTS=$(BASE_PROJECTS) $(FOREIGN_PROJECTS) $(STUB_PROJECTS)
 GENERATED=src/ctypes_config.h src/ctypes_config.ml setup.data src/ctypes/ctypes_primitives.ml src/ctypes-foreign-base/dl.ml src/ctypes-foreign-base/dl_stubs.c
 OCAML_FFI_INCOPTS=$(libffi_opt)
 export CFLAGS DEBUG
@@ -22,9 +25,11 @@ OS_ALT_SUFFIX=.unix
 endif
 
 # public targets
-all: setup.data build
+all: setup.data $(PROJECTS)
 
-build: $(PROJECTS)
+ctypes-base: $(BASE_PROJECTS)
+ctypes-foreign: ctypes-base $(FOREIGN_PROJECTS)
+ctypes-stubs: ctypes-base $(STUB_PROJECTS)
 
 clean:
 	rm -fr _build
@@ -168,7 +173,7 @@ install: META-install $(PROJECTS:%=install-%)
 uninstall:
 	$(OCAMLFIND) remove ctypes
 
-.PHONY: depend distclean clean build configure all install $(PROJECTS)
+.PHONY: depend distclean clean configure all install $(PROJECTS)
 
 include .depend Makefile.rules Makefile.examples Makefile.tests
 -include setup.data

--- a/Makefile
+++ b/Makefile
@@ -8,10 +8,15 @@ OCAMLMKLIB=$(OCAMLFIND) ocamlmklib
 VPATH=src examples
 BUILDDIR=_build
 BASE_PROJECTS=configure libffi-abigen configured ctypes ctypes-top
-FOREIGN_PROJECTS=ctypes-foreign-base ctypes-foreign-threaded ctypes-foreign-unthreaded
+FOREIGN_PROJECTS=test-libffi ctypes-foreign-base ctypes-foreign-threaded ctypes-foreign-unthreaded
 STUB_PROJECTS=cstubs
 PROJECTS=$(BASE_PROJECTS) $(FOREIGN_PROJECTS) $(STUB_PROJECTS)
-GENERATED=src/ctypes_config.h src/ctypes_config.ml libffi.config src/ctypes/ctypes_primitives.ml src/ctypes-foreign-base/dl.ml src/ctypes-foreign-base/dl_stubs.c
+GENERATED=src/ctypes/ctypes_primitives.ml	\
+          src/ctypes-foreign-base/libffi_abi.ml \
+          src/ctypes-foreign-base/dl.ml		\
+          src/ctypes-foreign-base/dl_stubs.c	\
+          libffi.config				\
+          asneeded.config 
 OCAML_FFI_INCOPTS=$(libffi_opt)
 export CFLAGS DEBUG
 
@@ -33,8 +38,6 @@ ctypes-stubs: ctypes-base $(STUB_PROJECTS)
 
 clean:
 	rm -fr _build
-
-distclean: clean
 	rm -f $(GENERATED)
 
 # ctypes subproject
@@ -175,8 +178,24 @@ install: META-install $(PROJECTS:%=install-%)
 uninstall:
 	$(OCAMLFIND) remove ctypes
 
-.PHONY: depend distclean clean configure all install $(PROJECTS)
+.PHONY: depend clean configure all install $(PROJECTS)
 
 include .depend Makefile.rules Makefile.examples Makefile.tests
 -include libffi.config
 -include asneeded.config
+
+ifeq ($(libffi_available),false)
+test-libffi:
+	@echo "The following required C libraries are missing: libffi."
+	@echo "Please install them and retry. If they are installed in a non-standard location"
+	@echo "or need special flags, set the environment variables <LIB>_CFLAGS and <LIB>_LIBS"
+	@echo "accordingly and retry."
+	@echo
+	@echo " For example, if libffi is installed in /opt/local, you can type:"
+	@echo
+	@echo "   export LIBFFI_CFLAGS=-I/opt/local/include"
+	@echo "   export LIBFFI_LIBS=-L/opt/local/lib"
+	@exit 1
+else:
+test-libffi:
+endif

--- a/appveyor/build.sh
+++ b/appveyor/build.sh
@@ -71,9 +71,9 @@ else
     export PKG_CONFIG_LIBDIR=${godi_dir}/lib/pkgconfig
 fi
 
-touch setup.data
+touch libffi.config
 make distclean || true
-rm -f setup.data
+rm -f libffi.config
 make all
 make date date-stubs date-stub-generator date-cmd-build date-cmd
 ./_build/date-cmd.native

--- a/src/discover/discover.ml
+++ b/src/discover/discover.ml
@@ -56,18 +56,6 @@ let ( // ) = Filename.concat
 (** See the comment in commands.ml *)
 let unixify = Str.(global_replace (regexp "\\") "/")
 
-let advice = "
-The following required C libraries are missing: libffi.
-Please install them and retry. If they are installed in a non-standard location
-or need special flags, set the environment variables <LIB>_CFLAGS and <LIB>_LIBS
-accordingly and retry.
-
-For example, if libffi is installed in /opt/local, you can type:
-
-export LIBFFI_CFLAGS=-I/opt/local/include
-export LIBFFI_LIBS=-L/opt/local/lib
-
-"
 (* +-----------------------------------------------------------------+
    | Test codes                                                      |
    +-----------------------------------------------------------------+ *)
@@ -223,7 +211,9 @@ let test_libffi setup_data have_pkg_config =
       | None    , None    , opt, lib -> opt, lib
   in
   setup_data := ("libffi_opt", opt) :: ("libffi_lib", lib) :: !setup_data;
-  test_code opt lib libffi_stub_code libffi_caml_code
+  let libffi_available = test_code opt lib libffi_stub_code libffi_caml_code in
+  setup_data := ("libffi_available", [string_of_bool libffi_available]) :: !setup_data;
+  libffi_available
 
 (* Test for pkg-config. If we are on MacOS X, we need the latest pkg-config
  * from Homebrew *)
@@ -267,10 +257,10 @@ let () =
       (fun () -> test_libffi setup_data have_pkg_config) in
 
   if not have_pkg_config then
-    fprintf stderr "Warning: the 'pkg-config' command is not available.";
+    fprintf stderr "Warning: the 'pkg-config' command is not available.\n";
 
   if not have_libffi then
-    (prerr_endline advice; exit 1);
+    fprintf stderr "Warning: libffi is not available.\n";
 
   ListLabels.iter !setup_data
     ~f:(fun (name, args) ->

